qiskit/transpiler/_transpiler.py

```diff
@@ -88,92 +79,83 @@ def compile(circuits, backend,
-            # we assume that compile_circuit has already expanded gates
-            # to the target basis, so we just need to generate json
-            json_circuit = DagUnroller(dag_circuit, JsonBackend(dag_circuit.basis)).execute()
-            job["compiled_circuit"] = json_circuit
-            # set eval_symbols=True to evaluate each symbolic expression
-            # TODO after transition to qobj, we can drop this
-            job["compiled_circuit_qasm"] = dag_circuit.qasm(qeflag=True,
-                                                            eval_symbols=True)
+        job["config"]["seed"] = seed
+
+        # step 3: populate the circuit `instructions` after compilation
+        # step 3a: circuit -> dag
+        dag_circuit = DAGCircuit.fromQuantumCircuit(circuit)
+
+        # TODO: move this inside the mapper pass
+        # pick a good initial layout if coupling_map is not already satisfied
+        # otherwise keep it as q[i]->q[i]
+        if (initial_layout is None and
+            coupling_map is not None and
+            not _matches_coupling_map(circuit.data, coupling_map)):
+            initial_layout = _pick_best_layout(backend, num_qubits, circuit.get_qregs())
+
+        # step 3b: transpile (dag -> dag)
+        dag_circuit, final_layout = transpile(
+            dag_circuit,
+            basis_gates=basis_gates,
+            coupling_map=coupling_map,
+            initial_layout=initial_layout,
+            get_layout=True,
+            pass_manager=pass_manager,
+            seed=seed)
+
+        # step 3c: dag -> json
+        # TODO: populate the Qobj object when Qobj class exists
+        # the compiled circuit to be run saved as a dag
+        # we assume that transpile() has already expanded gates
+        # to the target basis, so we just need to generate json
+        list_layout = [[k, v] for k, v in final_layout.items()] if final_layout else None
+        job["config"]["layout"] = list_layout
+        json_circuit = DagUnroller(dag_circuit, JsonBackend(dag_circuit.basis)).execute()
+        job["compiled_circuit"] = json_circuit
+
+        # set eval_symbols=True to evaluate each symbolic expression
+        # TODO after transition to qobj, we can drop this
+        job["compiled_circuit_qasm"] = dag_circuit.qasm(qeflag=True,
+                                                        eval_symbols=True)
+
         # add job to the qobj
         qobj["circuits"].append(job)
-    return qobj
 
+    return qobj
 
-def compile_circuit(quantum_circuit, basis_gates='u1,u2,u3,cx,id', coupling_map=None,
-                    initial_layout=None, get_layout=False, format='dag', seed=None):
-    """Compile the circuit.
 
-    This builds the internal "to execute" list which is list of quantum
-    circuits to run on different backends.
+def transpile(dag_circuit, basis_gates='u1,u2,u3,cx,id', coupling_map=None,
+              initial_layout=None, get_layout=False, pass_manager=None,
```

Comment:
Breaking change, please add the new `pass_manager` parameter at the end of the current parameter list.

Reply:
will do.